xds/src/main/java/io/grpc/xds/EdsLoadBalancer2.java

```diff
@@ -0,0 +1,421 @@
+    if (xdsClientPool == null) {
+      xdsClientPool = resolvedAddresses.getAttributes().get(XdsAttributes.XDS_CLIENT_POOL);
+      xdsClient = xdsClientPool.getObject();
+    }
+    EdsConfig config = (EdsConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+    if (logger.isLoggable(XdsLogLevel.INFO)) {
+      logger.log(XdsLogLevel.INFO, "Received EDS lb config: cluster={0}, "
+              + "eds_service_name={1}, endpoint_picking_policy={2}, report_load={3}",
+          config.clusterName, config.edsServiceName,
+          config.endpointPickingPolicy.getProvider().getPolicyName(),
+          config.lrsServerName != null);
+    }
+    if (cluster == null) {
+      cluster = config.clusterName;
+    }
+    if (edsLbState == null || !Objects.equals(edsLbState.edsServiceName, config.edsServiceName)) {
+      edsLbState = new EdsLbState(config.edsServiceName, config.lrsServerName);
+      switchingLoadBalancer.switchTo(edsLbState);
+    }
+    switchingLoadBalancer.handleResolvedAddresses(resolvedAddresses);
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    logger.log(XdsLogLevel.WARNING, "Received name resolution error: {0}", error);
+    switchingLoadBalancer.handleNameResolutionError(error);
+  }
+
+  @Override
+  public boolean canHandleEmptyAddressListFromNameResolution() {
+    return true;
+  }
+
+  @Override
+  public void shutdown() {
+    logger.log(XdsLogLevel.INFO, "Shutdown");
+    switchingLoadBalancer.shutdown();
+    if (xdsClientPool != null) {
+      xdsClientPool.returnObject(xdsClient);
+    }
+  }
+
+  private final class EdsLbState extends LoadBalancer.Factory {
+    @Nullable
+    private final String edsServiceName;
+    @Nullable
+    private final String lrsServerName;
+    private final String resourceName;
+
+    private EdsLbState(@Nullable String edsServiceName, @Nullable String lrsServerName) {
+      this.edsServiceName = edsServiceName;
+      this.lrsServerName = lrsServerName;
+      resourceName = edsServiceName == null ? cluster : edsServiceName;
+    }
+
+    @Override
+    public LoadBalancer newLoadBalancer(Helper helper) {
+      return new ChildLbState(helper);
+    }
+
+    private final class ChildLbState extends LoadBalancer implements EndpointWatcher {
+      @Nullable
+      private final LoadStatsStore loadStatsStore;
+      private final DropHandlingLbHelper lbHelper;
+      private List<EquivalentAddressGroup> endpointAddresses = Collections.emptyList();
+      private Map<Integer, Map<Locality, Integer>> prioritizedLocalityWeights
+          = Collections.emptyMap();
+      private PolicySelection localityPickingPolicy;
+      private PolicySelection endpointPickingPolicy;
+      @Nullable
+      private LoadBalancer lb;
+
+      private ChildLbState(Helper helper) {
+        if (lrsServerName != null) {
+          loadStatsStore = xdsClient.addClientStats(cluster, edsServiceName);
+          xdsClient.reportClientStats();
```

Comment:
Doesn't really matter. `addClientStats` creates the stats object for recording loads while `reportClientStats` creates the LRS RPC stream and performs load reporting. Since the LRS response handling is always serialized in the sync context, at the time sending the next load reporting request it always sees the newly created stats object for this cluster. So, the order of these two lines doesn't affect anything.